_posts/2019-06-05-quarkus-and-web-ui-development-mode.adoc

```diff
@@ -0,0 +1,110 @@
+
+In another terminal window go into the `webapp/` folder of the cloned project, and run:
+----
+$yarn proxy
+----
+This will start the Angular development server in proxy mode on port 4200. Now if you go to http://localhost:4200 you should see the same application as you saw on port 8080, but served by the Angular development server. If you make any changes to any of the Angular components set up in https://github.com/kabir/blog-quarkus-ui-development/blob/master/webapp/src/app/app.component.ts[`app.component.ts`] you will see your changes reflected.
+
+This is great!!! We can now be super-productive and make changes in both our front-end and our back-end, and see the changes reflected immediately without needing to rebuild, repackage and restart our application!
+
+==== Adjustments Summary
+
+The example project has already been set up with the needed adjustments to do this. These are the main things which need doing:
+
+. Make sure we handle the Angular routes in the UI rather than on the back-end. This is important for the bundled version (running on port 8080).
+. Enable the Angular proxy
+.. Signal to the back-end that the front-end is running in a proxy, i.e. not served by us. This is important for when the back-end needs to redirect to a page in the front-end.
+
+We will not go into details of how the application is set up, but will instead focus on the changes needed to make this work well.
+
+==== Handle Angular Routes
+We need to add a filter, https://github.com/kabir/blog-quarkus-ui-development/blob/master/src/main/java/org/kabir/quarkus/ui/AngularRouteFilter.java[AngularRouteFilter]. The essence of its `doFilter()` method is:
+[source,java]
+----
+chain.doFilter(request, response);
+
+if (response.getStatus() == 404) {
+    String path = request.getRequestURI().substring(
+            request.getContextPath().length()).replaceAll("[/]+$", "");
+    if (!FILE_NAME_PATTERN.matcher(path).matches()) {
+        // We could not find the resource, i.e. it is not anything known to the server (i.e. it is not a REST
+        // endpoint or a servlet), and does not look like a file so try handling it in the front-end routes.
+        request.getRequestDispatcher("/").forward(request, response);
+    }
+}
+
+----
+Without this filter, if you try to go to one of the Angular routes directly (e.g if you try to refresh when on http://localhost:8080/rest which is the `Rest` page we saw earlier) Quarkus will think it is a file, a REST endpoint or a servlet. Since it has no knowledge of anything called `/rest` you end up with a 404 (Not Found) status. `AngularRouteFilter` checks the status code, and if it is 404 and does not seem to be a file, we attempt to handle it in Angular.
```

Comment:
I understood that this filter will redirect localhost:8080/foo/bar to localhost:4200/foo/bar but either ti's not what it is doing precisely or I don't see how that redirection is configured.
Or I did not get the usefulness of the filter.

Reply:
No, this is only needed when running a packaged angular application, this is unrelated to the proxy.

Say you have an angular route called /other like in the example. If you load this from angular CLI it works ok (as angular knows what you are doing). If you are not familiar with routes in modern web applications, https://angular.io/guide/router might help.

However, when deployed in Quarkus (or any other app server for that matter) it will look for a resource called /other. There is no file called that, no REST endpoint and no servlet called that either. So the server will give a 404 error. 

All the server knows about from the angular side is that there is an index.html which pulls in the transpiled front-end resources. 

So in the case of something not found, we do a **forward** to the default location of the application, which loads that index.html and retains the path and query parameters to be parsed by Angular on the front-end. 

This allows us to bookmark pages. Without this you can only go to localhost:8080 (which loads up the index.html), but not directly to something handled by Angular like localhost:8080/other. This is the only time the filter is actually needed.

Reply:
I can try to explain a bit better in that section I guess